lib/pacer.go

```diff
@@ -0,0 +1,170 @@
+// Where:
+//   R = Instantaneous attack rate at elapsed time t, hits per nanosecond
+//   M = Mean attack rate over period P, sp.Mean, hits per nanosecond
+//   A = Amplitude of sine wave, sp.Amp, hits per nanosecond
+//   O = Offset of sine wave, sp.StartAt, radians
+//   P = Period of sine wave, sp.Period, nanoseconds
+//   t = Elapsed time since attack, nanoseconds
+// The attack rate (sp.HitsPerNs) is described by the equation:
+//
+// This equation is integrated with respect to time to derive the expected
+// number of hits served at time t after the attack began:
+//     H = Mt - (AP/2𝛑)cos(O+(2𝛑/P)t) + (AP/2𝛑)cos(O)
+// Where:
+//   H = Total number of hits triggered during t
+type SinePacer struct {
+	// The period of the sine wave, e.g. 20*time.Minute
+	// MUST BE > 0
+	Period time.Duration
+	// The mid-point of the sine wave in freq-per-Duration,
+	// MUST BE > 0
+	Mean Rate
+	// The amplitude of the sine wave in freq-per-Duration,
+	// MUST NOT BE EQUAL TO OR LARGER THAN MEAN
+	Amp Rate
+	// The offset, in radians, for the sine wave at t=0.
+	StartAt float64
+}
+
+var _ Pacer = SinePacer{}
+
+func (sp SinePacer) String() string {
+	return fmt.Sprintf("Sine{%s ± %s / %s, offset %g𝛑}", sp.Mean, sp.Amp, sp.Period, sp.StartAt/math.Pi)
+}
+
+// Pace returns the Duration until the next hit should be sent,
+// based on when the attack began and how many hits have been sent thus far.
+func (sp SinePacer) Pace(elapsedTime time.Duration, elapsedHits uint64) (time.Duration, bool) {
+	if sp.Period <= 0 || sp.Mean.HitsPerNs() <= 0 || sp.Amp.HitsPerNs() >= sp.Mean.HitsPerNs() {
+		// If the SinePacer configuration is invalid, stop the attack.
+		return 0, true
+	}
+	expectedHits := sp.hits(elapsedTime)
+	if elapsedHits < uint64(expectedHits) {
+		// Running behind, send next hit immediately.
+		return 0, false
+	}
+	// Re-arranging our hits equation to provide a duration given the number of
+	// requests sent is non-trivial, so we must solve for the duration numerically.
+	// math.Round() added here because we have to coerce to int64 nanoseconds
+	// at some point and it corrects a bunch of off-by-one problems.
+	nsPerHit := math.Round(1 / sp.HitsPerNs(elapsedTime))
+	hitsToWait := float64(elapsedHits+1) - expectedHits
+	nextHitIn := time.Duration(nsPerHit * hitsToWait)
+
+	// If we can't converge to an error of <1e-3 within 5 iterations, bail.
+	// This rarely even loops for any large Period if hitsToWait is small.
+	for i := 0; i < 5; i++ {
+		hitsAtGuess := sp.hits(elapsedTime + nextHitIn)
+		err := float64(elapsedHits+1) - hitsAtGuess
+		if math.Abs(err) < 1e-3 {
+			return nextHitIn, false
+		}
+		nextHitIn = time.Duration(float64(nextHitIn) / (hitsAtGuess - float64(elapsedHits)))
+	}
+	return nextHitIn, false
+}
+
+// AmpHits returns AP/2𝛑, which is the number of hits added or subtracted
+// from the Mean due to the Amplitude over a quarter of the Period,
+// i.e. from 0 → 𝛑/2 radians
+func (sp SinePacer) AmpHits() float64 {
```

Comment:
Do helper-methods like these need to be exported (starting with capital letter) ? If we need them for implementation only, then it would be clearer for the end user if we hide them from him.

Reply:
Good question.

Arguably not, I suppose. I exported them just in case they might be useful but I've not yet found them to be. I'll make them private.